torch/distributed/optim/zero_redundancy_optimizer.py

```diff
@@ -419,172 +448,186 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
-            # trainable params
-            buffer_size = 0
+            # Find the bucket size and dtype, compile the trainable
+            # parameters, and clone the non-trainable parameters
+            bucket_size = 0
             dtype = None
             trainable_params = []
             for param_group in param_groups:
                 for param in param_group["params"]:
                     if not _is_trainable(param):
+                        # Clone in case the parameter was previously part of
+                        # a bucket to avoid the data from being destroyed
                         param.data = param.data.detach().clone()
                     else:
-                        buffer_size += param.numel()
+                        bucket_size += param.numel()
                         trainable_params.append(param)
-                    dtype = param.dtype  # assumes all dense and same dtype
-
-            # Create a dummy bucket if there are no params
-            if buffer_size == 0:
-                self.buckets.append(torch.zeros(1, device=self._device))
-                continue
-
-            # Otherwise, construct the bucket
-            bucket = torch.empty(buffer_size, dtype=dtype, device=self._device)
-            offset = 0
-            for param in trainable_params:
-                offset_next = offset + param.numel()
-                bucket[offset:offset_next].copy_(param.data.flatten())
-                param.data = bucket[offset:offset_next].view_as(param.data)
-                offset = offset_next
+                    dtype = param.dtype  # assumes all same dtype
+            device = self._default_device  # assumes all on single device
+
+            if bucket_size == 0:
+                # Create a dummy bucket if there are no parameters
+                bucket = torch.zeros(1, device=device)
+            else:
+                # Construct the bucket (assuming all dense and same dtype)
+                bucket = torch.empty(bucket_size, dtype=dtype, device=device)
+                offset = 0
+                for param in trainable_params:
+                    offset_next = offset + param.numel()
+                    bucket[offset:offset_next].copy_(param.data.flatten())
+                    param.data = bucket[offset:offset_next].view_as(param.data)
+                    offset = offset_next
 
             # Either replace the existing bucket or create it
-            if len(self.buckets) != rank:
-                self.buckets[rank] = bucket
+            if len(self._buckets) != rank:
+                self._buckets[rank] = bucket
             else:
-                self.buckets.append(bucket)
+                self._buckets.append(bucket)
 
     def _update_trainable(self) -> None:
         r"""
         Updates the partitioning and communication patterns if the trainability
-        (``requires_grad``) of some parameters changed.
+        (``requires_grad``) of some parameters has changed.
```

Comment:
has -> have?

Reply:
Very minor, but I believe the current way is correct since it should be `trainability` [...] `has changed` rather than `trainability` [...] `have changed` (`trainability` being singular not plural)
:)

Reply:
Oh I see, my bad, thanks!